.github/workflows/main.yml

```diff
@@ -0,0 +1,108 @@
+name: Continuous Integration Tests
+
+on:
+  push:
+    branches:
+      - master 
+      - release
+  pull_request:
+    branches:
+      - master 
+      - release
+jobs:
+  lint:
+    strategy:
+      fail-fast: false #ensure both tests run even if one fails
+      matrix:
+        os: [ubuntu-latest, windows-latest]   
+    runs-on: ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up JDK 1.8
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+      - name: Cache Gradle packages
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.gradle/caches
+            ~/.gradle/wrapper
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*') }}
+          restore-keys: |
+            ${{ runner.os }}-gradle-
+      - name: Run Backend Linting 
+        run: ./gradlew lint --continue 
+      - name: Install Frontend Dependencies
+        run: npm ci 
+      - name: Run Lint Checks 
+        run: npm run lint 
+  component-testing:
+    strategy:
+      fail-fast: false #ensure both tests run even if one fails
+      matrix:
+        os: [ubuntu-latest, windows-latest]   
+    runs-on: ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up JDK 1.8
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+      - name: Cache Gradle packages
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.gradle/caches
+            ~/.gradle/wrapper
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*') }}
+          restore-keys: |
+            ${{ runner.os }}-gradle-
+      - name: Run Backend Tests    
+        run: |
+          ./gradlew createConfigs componentTests 
+          ./gradlew jacocoReport
+      - name: Generate Types 
+        run: ./gradlew generateTypes
+      - name: Install Frontend Dependencies 
+        run: npm ci 
+      - name: Run Frontend Tests 
+        run: npm run coverage 
+      - name: Upload to Codecov 
+        if: matrix.os == 'ubuntu-latest' #only upload to codecov on ubuntu VM 
+        run: bash <(curl -s https://codecov.io/bash)
+  E2E-testing:
+    runs-on: ubuntu-latest
```

Comment:
Any reason why the component and lint is run on both ubuntu and windows while e2e only on ubuntu? I thought it should be the other way around

Reply:
That is a good question - I followed our current configuration for this. We only run E2E tests on ubuntu at the moment.

Reply:
The reason could be that E2E tests are expensive. We don't want to waste build minutes even if they are free. I think Linux minutes are cheaper than Windows and Mac minutes.
If possible to configure as such, perhaps we can run E2E on multiple platforms only when a branch is merged to master and test only on one platform for PR branches.

Reply:
> If possible to configure as such, perhaps we can run E2E on multiple platforms only when a branch is merged to master and test only on one platform for PR branches.

Yes this is definitely something we can do with Github Actions! I can look into it once this PR is okay

Reply:
For PRs, we shouldn't do more than necessary because each run adds up to developer waiting time. In fact, all the things done on windows platform should not be a hard necessity for PRs. Running on more than one browser for PRs do not actually value-add much either, especially given that E2E tests are not expected to run often these days.